### PR TITLE
[WIP] Add `snapshot` and `update` for `TemporalData`

### PR DESCRIPTION
**Pull request by EdisonLeeeee**
Hi, PyG team.
I made a PR to implement tasks 5 and 7 in https://github.com/pyg-team/pytorch_geometric/issues/3230.

Based on your discussions, and the efforts made by @otaviocx, I made some changes to `TemporalData`. If I understood correctly, this PR should cover the main functionality of `snapshot` and `update` in `TemporalData`. 

Let me introduce my solution for implementing `snapshot` method. Motivated by graph-level datasets, I added an auxiliary variable `_slices`, which records the slices of each timed snapshot. We can easily get the snapshot by `_slices`, as like the implementation in `graph-level` datasets.

NOTE that this PR is a draft and there are some ambiguities:
+ `snapshot(start, end)`: what's the meaning of `start` and `end`? The orders of events or the exact timestamp of events? Currently I pick the former for simplicity.
+ `snapshot` should return `Data` or `TemporalData`? Currently I returned `Data` as your discussions.
+ `update`: how to deal with node evolving events (e.g., only `src` exists)? Treat them as self-loops  as suggested by @otaviocx? Currently I did nothing for this.

Excuse me if I made something wrong on this roadmap.

Here are some examples/tests for `TempralData` in this PR:

### `Initialization`

+ Initialize events `(src, dst, t, msg)`
```python
import torch
from torch_geometric.data import TemporalData


src = torch.tensor([1, 2, 3])
dst = torch.tensor([4, 5, 6])
t = torch.tensor([2022, 2023, 2024])
msg = torch.ones(src.size(0), 5)
```
+ `TemporalData` with full arguments
```python
data = TemporalData(src, dst, t, msg)
data
# TemporalData(src=[3], dst=[3], t=[3], msg=[3, 5])
```
+ `TemporalData` with missing argument `msg` 
```python

data = TemporalData(src, dst, t)
data
# TemporalData(src=[3], dst=[3], t=[3])
```

+  `TemporalData` with missing  arguments `t` and `msg`. For missing `t`, I let `t=0` for all events
```python
data = TemporalData(src, dst)
assert torch.all(data.t == 0)
data
# TemporalData(src=[3], dst=[3], t=[3])
```

+ `TemporalData` with only one argument `src`

```python
data = TemporalData(src)
assert torch.all(data.t == 0)
data
# TemporalData(src=[3], t=[3])

```

### `update`
+ `TemporalData` with single node evolving. if `t` is not specified, it is assigned `last_time+1` for all events
```python
data = TemporalData(src, dst, t, msg)
data.update(5) # add one node `5`
data
# TemporalData(src=[4], dst=[3], t=[4], msg=[3, 5])

```
+ `TemporalData` with single event evolving
```python
data = TemporalData(src, dst, t, msg)
# here `t` must be larger than last time in `data`
data.update(5, 10, t=2025) # add an event (5, 10, 2025)
data
# TemporalData(src=[4], dst=[4], t=[4], msg=[3, 5])
```

+ `TemporalData` with multiple events evolving
```python
data = TemporalData(src, dst, t, msg)
data.update(torch.arange(3), torch.arange(3)+1, t=torch.tensor([2025, 2025, 2025]))
data
# TemporalData(src=[6], dst=[6], t=[6], msg=[3, 5])

```

### `snapshot`
+ A snapshot starting from 1st time step
```python
data = TemporalData(src, dst, t, msg)
data.snapshot(start=1)
# Data(edge_index=[2, 2], edge_attr=[2, 5], time_attr=[2])

```
+ A snapshot ending at 3rd time step
```python
data = TemporalData(src, dst, t, msg)
data.snapshot(end=3)
# Data(edge_index=[2, 3], edge_attr=[3, 5], time_attr=[3])
```

+ A snapshot starting from 1st time step and ending at 3rd time step
```python
data = TemporalData(src, dst, t, msg)
data.snapshot(start=1, end=3)
# Data(edge_index=[2, 2], edge_attr=[2, 5], time_attr=[2])

```

+ The whole graph
```python
data = TemporalData(src, dst, t, msg)
data.snapshot()
# Data(edge_index=[2, 3], edge_attr=[3, 5], time_attr=[3])

```

It currently is a draft and requires more review and checks. It also has several potential bugs as I treated `dst` and `msg` as `None` by default. For example
```python
data.update(src, dst)
data.update(src)
data.update(src, dst)
# !!! ERROR. `src` and  `dst` mismatch!!
# this bug is also unavoidable for `msg`.
```

Any suggestions would be appreciated.

cc: @rusty1s

### CODE REVIEW

File: torch_geometric/data/temporal.py [hunk 1]
```diff
@@ -97,14 +98,74 @@ def __init__(
...
+        if t is None:
+            t = torch.full_like(src, 0)
+        else:
+            src, dst, t, msg = sort_events(src, dst, t, msg)
         self.src = src
         self.dst = dst
         self.t = t
         self.msg = msg
 
+        self._unique_time, inverse_indices, counts = torch.unique_consecutive(
```
Comment (Padarn) [inline torch_geometric/data/temporal.py:110]:
```suggestion
        self.unique_time, inverse_indices, counts = torch.unique_consecutive(
```
personal preference

File: torch_geometric/data/temporal.py [hunk 2]
```diff
@@ -207,6 +268,16 @@ def num_events(self) -> int:
         """
         return self.src.size(0)
 
+    @property
+    def num_snapshots(self) -> int:
```
Comment (Padarn) [inline torch_geometric/data/temporal.py:272]: do you think we should just expose `unique_times`? That way people also can see gaps between the snapshots?

**Review by Padarn — COMMENTED** [review]
Sorry haven't had time to look in heaps of detail, left a few small comments

File: torch_geometric/data/temporal.py [hunk 3]
```diff
@@ -97,14 +98,74 @@ def __init__(
...
+        :obj:`(src, dst, t, msg)`
+        """
+        src = torch.tensor(src).view(-1)
+
+        last_time = self._unique_time[-1]
+
+        if t is None:
+            t = torch.full_like(src, last_time + 1)
+        else:
+            t = torch.tensor(t).view(-1)
```
Comment (Padarn) [inline torch_geometric/data/temporal.py:144]: Why do we take the view? Can we assume that users will provide a certain input shape?

File: torch_geometric/data/temporal.py [hunk 4]
```diff
@@ -97,14 +98,74 @@ def __init__(
...
+            assert torch.all(t > last_time)
```
Comment (Padarn) [inline torch_geometric/data/temporal.py:145]: In this case we could rebuild the index entirely, but provide a warning? (or even have a 'allow_rebuild'). WDTY??

**Comment by EdisonLeeeee** [comment]
Sorry, I meant to make another new PR based on my branch and accidentally closed this one. I will reopen it if there is further discussion on this functionality. 

Thanks @Padarn again for the helpful comments.

**Pull request closed without merging**